File: hotalg/wish/L53.py
"""
给你一个整数数组 nums ，请你找出一个具有最大和的连续子数组（子数组最少包含一个元素），返回其最大和。

子数组
是数组中的一个连续部分。
"""
class Solution(object):
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        l = len(nums)
        sum = [0] * (l + 1)
        minnum = 0
        maxans = nums[0]
        for x, i in zip(nums, range(0, l)):
            sum[i+1] = sum[i] + x
            maxans = max(maxans, sum[i+1] - minnum)
            minnum = min(minnum, sum[i+1])
        return maxans

File: algorithm/src/main/java/sulwish/L53.java
package sulwish;

/*
给你一个整数数组 nums ，请你找出一个具有最大和的连续子数组（子数组最少包含一个元素），返回其最大和。

s[j+1] - s[i] = sum[a[i],...,a[j])

子数组
是数组中的一个连续部分。
 */

public class L53 {
    public int maxSubArray(int[] nums) {
        int n = nums.length;
        int ans = Integer.MIN_VALUE;
        int[] s = new int[n+1];
        s[0] = 0;
        int curMin = 0;
        for (int i = 1; i <= n; i++){
            s[i] = s[i-1] + nums[i-1];
            ans = Math.max(ans, s[i] - curMin);
            curMin = Math.min(s[i], curMin);
        }
        return ans;
    }

    public static void main(String[] args) {
        System.out.println(new L53().maxSubArray(new int[]{-1}));
    }
}
